File: DotNET/Endpoint Examples/JSON Payload/signed-pdf.cs

/*
 * What this sample does:
 * - Called from Program.cs to digitally sign a PDF using a PFX, passphrase, and optional logo via JSON flow (uploads all inputs).
 *
 * Setup (environment):
 * - Copy .env.example to .env
 * - Set PDFREST_API_KEY=your_api_key_here
 * - Optional: set PDFREST_URL to override the API region. For EU/GDPR compliance and proximity, use:
 *     PDFREST_URL=https://eu-api.pdfrest.com
 *   For more information visit https://pdfrest.com/pricing#how-do-eu-gdpr-api-calls-work
 *
 * Usage:
 *   dotnet run -- signed-pdf input.pdf creds.pfx passphrase.txt logo.png
 *
 * Output:
 * - Prints JSON responses; non-2xx results exit non-zero.
 */
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System.Text;

namespace Samples.EndpointExamples.JsonPayload
{
    public static class SignedPdf
    {
        public static async Task Execute(string[] args)
        {
            if (args == null || args.Length < 4)
            {
                Console.Error.WriteLine("signed-pdf requires <input.pdf> <credentials.pfx> <passphrase.txt> <logo.png>");
                Environment.Exit(1);
                return;
            }

            var apiKey = Environment.GetEnvironmentVariable("PDFREST_API_KEY");
            if (string.IsNullOrWhiteSpace(apiKey)) { Console.Error.WriteLine("Missing required environment variable: PDFREST_API_KEY"); Environment.Exit(1); return; }
            var baseUrl = Environment.GetEnvironmentVariable("PDFREST_URL") ?? "https://api.pdfrest.com";
            var files = args.Take(4).ToArray();
            foreach (var f in files) { if (!File.Exists(f)) { Console.Error.WriteLine($"File not found: {f}"); Environment.Exit(1); return; } }

            using var httpClient = new HttpClient { BaseAddress = new Uri(baseUrl) };
            var ids = new string[files.Length];
            for (int i = 0; i < files.Length; i++)
            {
                var uploadRequest = new HttpRequestMessage(HttpMethod.Post, "upload");
                uploadRequest.Headers.TryAddWithoutValidation("Api-Key", apiKey);
                uploadRequest.Headers.Accept.Add(new("application/json"));
                var bytes = File.ReadAllBytes(files[i]);
                var content = new ByteArrayContent(bytes);
                content.Headers.TryAddWithoutValidation("Content-Type", "application/octet-stream");
                content.Headers.TryAddWithoutValidation("Content-Filename", Path.GetFileName(files[i]));
                uploadRequest.Content = content;
                var uploadResponse = await httpClient.SendAsync(uploadRequest);
                var uploadResult = await uploadResponse.Content.ReadAsStringAsync();
                JObject uploadJson = JObject.Parse(uploadResult);
                ids[i] = uploadJson["files"][0]["id"].ToString();
            }

            var signedPdfRequest = new HttpRequestMessage(HttpMethod.Post, "signed-pdf");
            signedPdfRequest.Headers.TryAddWithoutValidation("Api-Key", apiKey);
            signedPdfRequest.Headers.Accept.Add(new("application/json"));
            signedPdfRequest.Headers.TryAddWithoutValidation("Content-Type", "application/json");

            var signatureConfiguration = new JObject
            {
                ["type"] = "new",
                ["name"] = "esignature",
                ["logo_opacity"] = "0.5",
                ["location"] = new JObject
                {
                    ["bottom_left"] = new JObject { ["x"] = "0", ["y"] = "0" },
                    ["top_right"] = new JObject { ["x"] = "216", ["y"] = "72" },
                    ["page"] = "1"
                },
                ["display"] = new JObject
                {
                    ["include_distinguished_name"] = "true",
                    ["include_datetime"] = "true",
                    ["contact"] = "My contact information",
                    ["location"] = "My signing location",
                    ["name"] = "John Doe",
                    ["reason"] = "My reason for signing"
                }
            };

            JObject parameterJson = new JObject
            {
                ["id"] = ids[0],
                ["pfx_credential_id"] = ids[1],
                ["pfx_passphrase_id"] = ids[2],
                ["logo_id"] = ids[3],
                ["signature_configuration"] = signatureConfiguration.ToString(Formatting.None),
            };
            signedPdfRequest.Content = new StringContent(parameterJson.ToString(), Encoding.UTF8, "application/json");
            var signedPdfResponse = await httpClient.SendAsync(signedPdfRequest);
            var signedPdfResult = await signedPdfResponse.Content.ReadAsStringAsync();
            Console.WriteLine("Processing response received.");
            Console.WriteLine(signedPdfResult);
        }
    }
}

File: Java/Endpoint Examples/JSON Payload/SignedPDF.java
import io.github.cdimascio.dotenv.Dotenv;
import java.io.File;
import java.io.IOException;
import java.util.concurrent.TimeUnit;
import okhttp3.*;
import org.json.JSONArray;
import org.json.JSONObject;

public class SignedPDF {

    // Specify your API key here, or in the environment variable PDFREST_API_KEY.
    // You can also put the environment variable in a .env file.
    private static final String DEFAULT_API_KEY = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx";

    public static void main(String[] args) {
        File inputFile, credentialFile, passphraseFile, logoFile;
        if (args.length > 3) {
            inputFile = new File(args[0]);
            credentialFile = new File(args[1]);
            passphraseFile = new File(args[2]);
            logoFile = new File(args[3]);
        } else {
            inputFile = new File("/path/to/input.pdf");
            credentialFile = new File("/path/to/credentials.pfx");
            passphraseFile = new File("/path/to/passphrase.txt");
            logoFile = new File("/path/to/logo.png");
        }
        final Dotenv dotenv = Dotenv.configure().ignoreIfMalformed().ignoreIfMissing().load();

        String inputId = uploadFile(inputFile);
        String credentialId = uploadFile(credentialFile);
        String passphraseId = uploadFile(passphraseFile);
        String logoId = uploadFile(logoFile);

        JSONObject bottomLeft = new JSONObject();
        bottomLeft.put("x", "0");
        bottomLeft.put("y", "0");

        JSONObject topRight = new JSONObject();
        topRight.put("x", "216");
        topRight.put("y", "72");

        JSONObject location = new JSONObject();
        location.put("bottom_left", bottomLeft);
        location.put("top_right", topRight);
        location.put("page", "1");

        JSONObject display = new JSONObject();
        display.put("include_distinguished_name", "true");
        display.put("include_datetime", "true");
        display.put("contact", "My contact information");
        display.put("location", "My signing location");
        display.put("name", "John Doe");
        display.put("reason", "My reason for signing");

        JSONObject signatureConfig = new JSONObject();
        signatureConfig.put("type", "new");
        signatureConfig.put("name", "esignature");
        signatureConfig.put("logo_opacity", "0.5");
        signatureConfig.put("location", location);
        signatureConfig.put("display", display);

        JSONObject requestJson = new JSONObject();
        requestJson.put("id", inputId);
        requestJson.put("pfx_credential_id", credentialId);
        requestJson.put("pfx_passphrase_id", passphraseId);
        requestJson.put("logo_id", logoId);
        requestJson.put("signature_configuration", signatureConfig.toString());

        final RequestBody requestBody =
                RequestBody.create(requestJson.toString(), MediaType.parse("application/json"));

        Request request =
                new Request.Builder()
                        .header("Api-Key", dotenv.get("PDFREST_API_KEY", DEFAULT_API_KEY))
                        .url("https://api.pdfrest.com/signed-pdf")
                        .post(requestBody)
                        .build();
        try {
            OkHttpClient client =
                    new OkHttpClient().newBuilder().readTimeout(60, TimeUnit.SECONDS).build();

            Response response = client.newCall(request).execute();
            System.out.println("Processing Result code " + response.code());
            if (response.body() != null) {
                System.out.println(prettyJson(response.body().string()));
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    private static String prettyJson(String json) {
        // https://stackoverflow.com/a/9583835/11996393
        return new JSONObject(json).toString(4);
    }

    private static String uploadFile(File inputFile) {

        final Dotenv dotenv = Dotenv.configure().ignoreIfMalformed().ignoreIfMissing().load();

        final RequestBody requestBody =
                RequestBody.create(inputFile, MediaType.parse("application/pdf"));

        Request request =
                new Request.Builder()
                        .header("Api-Key", dotenv.get("PDFREST_API_KEY", DEFAULT_API_KEY))
                        .header("Content-Filename", inputFile.getName())
                        .url("https://api.pdfrest.com/upload")
                        .post(requestBody)
                        .build();
        try {
            OkHttpClient client = new OkHttpClient().newBuilder().build();
            Response response = client.newCall(request).execute();
            System.out.println("Upload Result code " + response.code());
            if (response.body() != null) {
                String inputUploadString = response.body().string();
                JSONObject inputUploadJSON = new JSONObject(inputUploadString);
                if (inputUploadJSON.has("error")) {
                    System.out.println("Error during upload: " + inputUploadString);
                    return "";
                }
                JSONArray fileArray = inputUploadJSON.getJSONArray("files");

                JSONObject fileObject = fileArray.getJSONObject(0);

                return fileObject.get("id").toString();
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        return "";
    }
}
